In [3]:
from pathlib import Path

LST_DIR = Path("../data_raw/LST/extracted")

lst_files = sorted(LST_DIR.glob("*.nc"))

print("Number of LST files:", len(lst_files))

for file in lst_files:
    print(file.name)

Number of LST files: 12
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20210601000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20210701000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20210801000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20220601000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20220701000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20220801000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20230601000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20230701000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20230801000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI-LST-L3S-LST-IRCDR_-0.01deg_1MONTHLY_DAY-20240601000000-fv3.00.area-subset.53.25.37.-10.nc
ESACCI

In [3]:
pip install xarray netCDF4

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 10.2 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 22.8/22.8 MB 50.5 MB/s  0:00:00 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3/3 [xarray]2m2/3 [xarray]]
Note: you may need to restart the kernel to use updated packages.


In [4]:
import xarray as xr

sample = xr.open_dataset(lst_files[0], engine="h5netcdf")

print(sample)

<xarray.Dataset> Size: 291MB
Dimensions:          (time: 1, lat: 1600, lon: 3500, length_scale: 1, channel: 2)
Coordinates:
  * time             (time) datetime64[ns] 8B 2021-06-01
  * lat              (lat) float32 6kB 37.01 37.01 37.02 ... 52.98 52.98 52.99
  * lon              (lon) float32 14kB -9.995 -9.985 -9.975 ... 24.98 24.99
  * channel          (channel) float32 8B 11.0 12.0
Dimensions without coordinates: length_scale
Data variables: (12/14)
    dtime            (time, lat, lon) float32 22MB ...
    satze            (time, lat, lon) float32 22MB ...
    sataz            (time, lat, lon) float32 22MB ...
    solze            (time, lat, lon) float32 22MB ...
    solaz            (time, lat, lon) float32 22MB ...
    lst              (time, lat, lon) float32 22MB ...
    ...               ...
    lst_unc_loc_atm  (time, lat, lon) float32 22MB ...
    lst_unc_loc_sfc  (time, lat, lon) float32 22MB ...
    lst_unc_sys      (length_scale) float32 4B ...
    lcc              (tim

In [5]:
print(sample["lst"])

<xarray.DataArray 'lst' (time: 1, lat: 1600, lon: 3500)> Size: 22MB
[5600000 values with dtype=float32]
Coordinates:
  * time     (time) datetime64[ns] 8B 2021-06-01
  * lat      (lat) float32 6kB 37.01 37.01 37.02 37.03 ... 52.98 52.98 52.99
  * lon      (lon) float32 14kB -9.995 -9.985 -9.975 ... 24.98 24.98 24.99
Attributes:
    long_name:     land surface temperature
    units:         kelvin
    valid_min:     -8315
    valid_max:     7685
    actual_range:  [190.04999 342.71   ]


In [6]:
lst = sample["lst"]

print("Min LST (K):", float(lst.min(skipna=True)))
print("Max LST (K):", float(lst.max(skipna=True)))
print("Mean LST (K):", float(lst.mean(skipna=True)))

Min LST (K): 190.97999572753906
Max LST (K): 326.8499755859375
Mean LST (K): 303.6390380859375


In [7]:
lst_c = sample["lst"] - 273.15
lst_c.attrs["units"] = "°C"

print("Min LST (°C):", float(lst_c.min(skipna=True)))
print("Max LST (°C):", float(lst_c.max(skipna=True)))
print("Mean LST (°C):", float(lst_c.mean(skipna=True)))

Min LST (°C): -82.16999816894531
Max LST (°C): 53.699981689453125
Mean LST (°C): 30.489049911499023


In [8]:
print(list(sample.data_vars))

['dtime', 'satze', 'sataz', 'solze', 'solaz', 'lst', 'lst_uncertainty', 'lst_unc_ran', 'lst_unc_loc_atm', 'lst_unc_loc_sfc', 'lst_unc_sys', 'lcc', 'n', 'lst_unc_loc_cor']


In [9]:
print("n:")
print(sample["n"].attrs)

print("\nlcc:")
print(sample["lcc"].attrs)

print("\nlst_uncertainty:")
print(sample["lst_uncertainty"].attrs)

n:
{'long_name': 'number of clear-sky pixels', 'valid_min': np.int16(0), 'valid_max': np.int16(18750), 'actual_range': array([ 1, 24], dtype=int16)}

lcc:
{'long_name': 'land cover class', 'units': '1', 'flag_meanings': 'cropland_rainfed cropland_rainfed_herbaceous_cover cropland_rainfed_tree_or_shrub_cover cropland_irrigated mosaic_cropland mosaic_natural_vegetation tree_broadleaved_evergreen_closed_to_open tree_broadleaved_deciduous_closed_to_open tree_broadleaved_deciduous_closed tree_broadleaved_deciduous_open tree_needleleaved_evergreen_closed_to_open tree_needleleaved_evergreen_closed tree_needleleaved_evergreen_open tree_needleleaved_deciduous_closed_to_open tree_needleleaved_deciduous_closed tree_needleleaved_deciduous_open tree_mixed mosaic_tree_and_shrub mosaic_herbaceous shrubland shrubland_evergreen shrubland_deciduous grassland lichens_and_mosses sparse_vegetation sparse_tree sparse_shrub sparse_herbaceous tree_cover_flooded_fresh_or_brakish_water tree_cover_flooded_saline

In [10]:
meanings = sample["lcc"].attrs["flag_meanings"].split()
values = sample["lcc"].attrs["flag_values"]

for value, meaning in zip(values, meanings):
    if meaning == "urban":
        print("Urban class value:", value)

Urban class value: 190


In [11]:
cities = {
    "Amsterdam": (52.3676, 4.9041),
    "Athens": (37.9838, 23.7275),
    "Barcelona": (41.3874, 2.1686),
    "Berlin": (52.5200, 13.4050),
    "Budapest": (47.4979, 19.0402),
    "Lisbon": (38.7223, -9.1393),
    "Madrid": (40.4168, -3.7038),
    "Milan": (45.4642, 9.1900),
    "Prague": (50.0755, 14.4378),
    "Vienna": (48.2082, 16.3738),
}

print("Number of cities:", len(cities))

Number of cities: 10


In [12]:
city = "Berlin"
lat, lon = cities[city]

berlin_pixel = sample["lst"].sel(
    lat=lat,
    lon=lon,
    method="nearest"
)

berlin_c = berlin_pixel - 273.15

print("City:", city)
print("LST (°C):", float(berlin_c.values.squeeze()))

City: Berlin
LST (°C): 38.660003662109375


In [13]:
berlin_lcc = sample["lcc"].sel(
    lat=lat,
    lon=lon,
    method="nearest"
)

print("Berlin land cover class:", int(berlin_lcc.values.squeeze()))

Berlin land cover class: 190


In [14]:
lat_buffer = 0.1
lon_buffer = 0.1

berlin_area = sample.sel(
    lat=slice(lat - lat_buffer, lat + lat_buffer),
    lon=slice(lon - lon_buffer, lon + lon_buffer)
)

urban_mask = berlin_area["lcc"] == 190
berlin_urban_lst = berlin_area["lst"].where(urban_mask) - 273.15

print("Urban pixels:", int(berlin_urban_lst.count()))
print("Mean urban LST (°C):", float(berlin_urban_lst.mean(skipna=True)))

Urban pixels: 382
Mean urban LST (°C): 37.086856842041016


In [15]:
import geopandas as gpd

berlin_file = list(
    Path("../data_raw/urban_atlas").rglob("*BERLIN*.fgb")
)[0]

berlin_ua = gpd.read_file(berlin_file)

print("CRS:", berlin_ua.crs)
print("Number of polygons:", len(berlin_ua))
print(berlin_ua.columns.tolist())

CRS: EPSG:3035
Number of polygons: 82024
['country', 'fua_name', 'fua_code', 'code_2021', 'class_2021', 'prod_date', 'identifier', 'perimeter', 'area', 'comment', 'geometry']


In [16]:
berlin_boundary = berlin_ua.dissolve()

berlin_boundary = berlin_boundary.to_crs("EPSG:4326")

print("CRS:", berlin_boundary.crs)
print("Bounds:", berlin_boundary.total_bounds)

CRS: EPSG:4326
Bounds: [12.434251  51.9792745 14.258758  53.172832 ]


## City-Level LST Extraction Using Urban Atlas Boundaries

To obtain representative city-level land surface temperatures, the Urban Atlas Functional Urban Area (FUA) boundaries are used instead of a single city-centre pixel or a fixed buffer.

The Urban Atlas boundaries are transformed from EPSG:3035 to EPSG:4326 to match the coordinate system of the LST dataset.

LST pixels located within each city boundary are then selected and aggregated to calculate the mean daytime land surface temperature.

This approach provides a consistent spatial basis for comparing summer LST across the 10 cities from 2021 to 2024.

In [17]:
from shapely.geometry import Point
import numpy as np

# Crop first to Berlin bounding box for speed
minx, miny, maxx, maxy = berlin_boundary.total_bounds

berlin_crop = sample.sel(
    lat=slice(miny, maxy),
    lon=slice(minx, maxx)
)

# Create lon/lat grid
lon2d, lat2d = np.meshgrid(
    berlin_crop["lon"].values,
    berlin_crop["lat"].values
)

# Berlin geometry
berlin_geom = berlin_boundary.geometry.iloc[0]

# Mask pixels inside Berlin boundary
mask = np.array([
    berlin_geom.contains(Point(x, y))
    for x, y in zip(lon2d.ravel(), lat2d.ravel())
]).reshape(lon2d.shape)

berlin_lst_city = (berlin_crop["lst"] - 273.15).where(mask)

print("Pixels inside Berlin:", int(berlin_lst_city.count()))
print("Mean LST Berlin (°C):", float(berlin_lst_city.mean(skipna=True)))

Pixels inside Berlin: 11705
Mean LST Berlin (°C): 32.48839569091797


## Urban Atlas Boundaries for the 10 Cities

Urban Atlas 2021 Functional Urban Area (FUA) data are used to define a consistent spatial boundary for each city.

The following step identifies the corresponding Urban Atlas file for each of the 10 cities before extracting city-level LST.

In [18]:
ua_names = {
    "Amsterdam": "AMSTERDAM",
    "Athens": "ATHINA",
    "Barcelona": "BARCELONA",
    "Berlin": "BERLIN",
    "Budapest": "BUDAPEST",
    "Lisbon": "LISBOA",
    "Madrid": "MADRID",
    "Milan": "MILANO",
    "Prague": "PRAHA",
    "Vienna": "WIEN",
}

urban_atlas_dir = Path("../data_raw/urban_atlas")

for city, ua_name in ua_names.items():
    matches = list(urban_atlas_dir.rglob(f"*{ua_name}*.fgb"))
    print(city, ":", len(matches))

Amsterdam : 1
Athens : 1
Barcelona : 1
Berlin : 1
Budapest : 1
Lisbon : 1
Madrid : 1
Milan : 1
Prague : 1
Vienna : 1


## Testing City-Level LST Extraction

Before processing the complete 2021–2024 dataset, the extraction method is tested for all 10 cities using June 2021.

For each city, the Urban Atlas FUA boundary is used to select LST pixels located within the city area. The mean daytime LST is then calculated in degrees Celsius.

This step validates the spatial extraction method before applying it to all summer months and years.

In [19]:
results_test = []

for city, ua_name in ua_names.items():

    # Load Urban Atlas boundary
    ua_file = list(
        urban_atlas_dir.rglob(f"*{ua_name}*.fgb")
    )[0]

    city_ua = gpd.read_file(ua_file)

    # Create one FUA boundary and convert to LST CRS
    city_boundary = city_ua.dissolve().to_crs("EPSG:4326")
    city_geom = city_boundary.geometry.iloc[0]

    # Crop LST to city bounding box
    minx, miny, maxx, maxy = city_boundary.total_bounds

    city_crop = sample.sel(
        lat=slice(miny, maxy),
        lon=slice(minx, maxx)
    )

    # Create pixel coordinates
    lon2d, lat2d = np.meshgrid(
        city_crop["lon"].values,
        city_crop["lat"].values
    )

    # Pixels inside city boundary
    mask = np.array([
        city_geom.contains(Point(x, y))
        for x, y in zip(lon2d.ravel(), lat2d.ravel())
    ]).reshape(lon2d.shape)

    # Mean LST in Celsius
    city_lst = (city_crop["lst"] - 273.15).where(mask)

    results_test.append({
        "city": city,
        "pixels": int(city_lst.count()),
        "mean_lst_c": float(city_lst.mean(skipna=True))
    })

results_test

/opt/miniconda3/lib/python3.13/site-packages/shapely/set_operations.py:553: RuntimeWarning: invalid value encountered in unary_union
  return lib.unary_union(collections, **kwargs)
/opt/miniconda3/lib/python3.13/site-packages/shapely/set_operations.py:553: RuntimeWarning: invalid value encountered in unary_union
  return lib.unary_union(collections, **kwargs)


[{'city': 'Amsterdam', 'pixels': 3361, 'mean_lst_c': 28.816776275634766},
 {'city': 'Athens', 'pixels': 1927, 'mean_lst_c': 38.4897346496582},
 {'city': 'Barcelona', 'pixels': 2811, 'mean_lst_c': 35.642024993896484},
 {'city': 'Berlin', 'pixels': 11705, 'mean_lst_c': 32.48839569091797},
 {'city': 'Budapest', 'pixels': 7560, 'mean_lst_c': 34.22037124633789},
 {'city': 'Lisbon', 'pixels': 4336, 'mean_lst_c': 35.2984733581543},
 {'city': 'Madrid', 'pixels': 8342, 'mean_lst_c': 38.75169372558594},
 {'city': 'Milan', 'pixels': 3594, 'mean_lst_c': 35.988807678222656},
 {'city': 'Prague', 'pixels': 7198, 'mean_lst_c': 30.45098114013672},
 {'city': 'Vienna', 'pixels': 11472, 'mean_lst_c': 33.548885345458984}]

## Summer LST Extraction: 2021–2024

The validated spatial extraction method is applied to all summer LST datasets from 2021 to 2024.

For each city and month, daytime LST pixels within the Urban Atlas FUA boundary are extracted and aggregated.

The resulting dataset contains monthly city-level LST values for June, July and August across four years, providing a temporal layer for comparing summer surface heat patterns between cities.

In [20]:
import pandas as pd
import re

# Prepare city boundaries once
city_boundaries = {}

for city, ua_name in ua_names.items():
    ua_file = list(
        urban_atlas_dir.rglob(f"*{ua_name}*.fgb")
    )[0]

    city_ua = gpd.read_file(ua_file)
    city_boundary = city_ua.dissolve().to_crs("EPSG:4326")

    city_boundaries[city] = city_boundary.geometry.iloc[0]


# Extract LST for all cities and all monthly files
lst_results = []

for file in lst_files:

    # Extract year and month from filename
    match = re.search(r"(\d{6})", file.name)
    date_code = match.group(1)

    year = int(date_code[:4])
    month = int(date_code[4:6])

    ds = xr.open_dataset(file, engine="h5netcdf")

    print(f"Processing {year}-{month:02d}...")

    for city, city_geom in city_boundaries.items():

        minx, miny, maxx, maxy = city_geom.bounds

        city_crop = ds.sel(
            lat=slice(miny, maxy),
            lon=slice(minx, maxx)
        )

        lon2d, lat2d = np.meshgrid(
            city_crop["lon"].values,
            city_crop["lat"].values
        )

        mask = np.array([
            city_geom.contains(Point(x, y))
            for x, y in zip(lon2d.ravel(), lat2d.ravel())
        ]).reshape(lon2d.shape)

        city_lst = (city_crop["lst"] - 273.15).where(mask)

        lst_results.append({
            "city": city,
            "year": year,
            "month": month,
            "pixel_count": int(city_lst.count()),
            "mean_lst_c": float(city_lst.mean(skipna=True))
        })

    ds.close()


lst_monthly = pd.DataFrame(lst_results)

print("Shape:", lst_monthly.shape)
lst_monthly.head(15)

/opt/miniconda3/lib/python3.13/site-packages/shapely/set_operations.py:553: RuntimeWarning: invalid value encountered in unary_union
  return lib.unary_union(collections, **kwargs)
/opt/miniconda3/lib/python3.13/site-packages/shapely/set_operations.py:553: RuntimeWarning: invalid value encountered in unary_union
  return lib.unary_union(collections, **kwargs)


Processing 2021-06...
Processing 2021-07...
Processing 2021-08...
Processing 2022-06...
Processing 2022-07...
Processing 2022-08...
Processing 2023-06...
Processing 2023-07...
Processing 2023-08...
Processing 2024-06...
Processing 2024-07...
Processing 2024-08...
Shape: (120, 5)


,city,year,month,pixel_count,mean_lst_c
0,Amsterdam,2021,6,3361,28.816776
1,Athens,2021,6,1927,38.489735
2,Barcelona,2021,6,2811,35.642025
3,Berlin,2021,6,11705,32.488396
4,Budapest,2021,6,7560,34.220371
5,Lisbon,2021,6,4336,35.298473
6,Madrid,2021,6,8342,38.751694
7,Milan,2021,6,3594,35.988808
8,Prague,2021,6,7198,30.450981
9,Vienna,2021,6,11472,33.548885


In [21]:
print("Shape:", lst_monthly.shape)
print("\nMissing values:")
print(lst_monthly.isna().sum())

print("\nPixel count range:")
print(lst_monthly["pixel_count"].describe())

Shape: (120, 5)

Missing values:
city           0
year           0
month          0
pixel_count    0
mean_lst_c     0
dtype: int64

Pixel count range:
count      120.000000
mean      6204.683333
std       3361.271448
min       1909.000000
25%       3347.750000
50%       5555.000000
75%       8334.000000
max      11706.000000
Name: pixel_count, dtype: float64


## Annual Summer LST Summary

The monthly LST values for June, July and August are aggregated to calculate the mean summer land surface temperature for each city and year.

This creates one summer LST value per city for each year from 2021 to 2024, allowing year-to-year comparison of summer surface heat patterns.

In [22]:
summer_lst = (
    lst_monthly
    .groupby(["city", "year"], as_index=False)
    .agg(
        summer_mean_lst_c=("mean_lst_c", "mean"),
        avg_pixel_count=("pixel_count", "mean")
    )
)

print("Shape:", summer_lst.shape)
summer_lst

Shape: (40, 4)


,city,year,summer_mean_lst_c,avg_pixel_count
0,Amsterdam,2021,26.624178,3305.000000
1,Amsterdam,2022,29.644736,3354.000000
2,Amsterdam,2023,27.445384,3342.333333
3,Amsterdam,2024,27.227859,3346.666667
4,Athens,2021,40.030468,1926.333333
5,Athens,2022,38.977674,1919.000000
6,Athens,2023,40.663328,1920.666667
7,Athens,2024,42.215510,1928.000000
8,Barcelona,2021,35.913541,2800.000000
9,Barcelona,2022,38.225863,2806.333333


## Summer LST Change Analysis: 2021–2024

To examine how summer surface heat changed across the study period, the annual summer mean LST values are compared between 2021 and 2024.

The analysis calculates the absolute temperature difference for each city and also identifies the warmest and coolest summer observed during the four-year period.

Because the dataset covers only four summers, the results are interpreted as recent temporal patterns rather than long-term climate trends.

In [23]:
lst_change = (
    summer_lst
    .pivot(index="city", columns="year", values="summer_mean_lst_c")
    .reset_index()
)

lst_change["change_2021_2024_c"] = (
    lst_change[2024] - lst_change[2021]
)

lst_change["warmest_year"] = (
    lst_change[[2021, 2022, 2023, 2024]]
    .idxmax(axis=1)
)

lst_change["coolest_year"] = (
    lst_change[[2021, 2022, 2023, 2024]]
    .idxmin(axis=1)
)

lst_change = lst_change.sort_values(
    "change_2021_2024_c",
    ascending=False
)

lst_change

year,city,2021,2022,2023,2024,change_2021_2024_c,warmest_year,coolest_year
8,Prague,28.538186,28.549192,30.843977,30.796501,2.258315,2023,2021
1,Athens,40.030468,38.977674,40.663328,42.215510,2.185042,2024,2022
9,Vienna,31.192992,32.458422,32.160390,32.858674,1.665682,2024,2021
4,Budapest,33.045675,37.580811,33.486053,34.692179,1.646503,2022,2021
6,Madrid,41.543050,44.091009,43.740168,42.726967,1.183917,2022,2021
0,Amsterdam,26.624178,29.644736,27.445384,27.227859,0.603682,2022,2021
5,Lisbon,36.671122,38.786237,38.917975,37.169290,0.498168,2023,2021
7,Milan,34.144988,37.885319,34.244956,34.480623,0.335635,2022,2021
3,Berlin,29.416946,30.920294,28.987234,29.558530,0.141584,2022,2023
2,Barcelona,35.913541,38.225863,37.309181,35.530212,-0.383329,2022,2024


## Linking Recent Summer LST with the Urban Cooling Priority Framework

The temporal LST results are compared with the existing Priority Score and UHI intensity to examine whether cities identified as higher cooling priorities also experience higher recent summer surface temperatures.

This analysis provides an additional heat perspective without incorporating LST directly into the Priority Score.

In [24]:
from pathlib import Path
import pandas as pd

final_path = Path("../Data_Processed/Smart_Cities_Cooler_Futures_Tableau_CLEAN.csv")

final_df = pd.read_csv(final_path)

print(final_df.shape)
print(final_df.columns.tolist())
print(final_df["city"].tolist())

(10, 21)
['city', 'building_density_per_km2', 'average_building_height_m', 'street_tree_pct', 'green_space_pct', 'uhi_intensity_c', 'population_density_per_km2', 'building_density_per_km2_score', 'average_building_height_m_score', 'street_tree_pct_score', 'green_space_pct_score', 'uhi_intensity_c_score', 'population_density_per_km2_score', 'priority_score', 'priority_rank', 'primary_driver', 'secondary_driver', 'recommended_intervention', 'impervious_increase_pct', 'impervious_decrease_pct', 'stable_impervious_pct']
['ATHINA', 'MILANO', 'AMSTERDAM', 'BARCELONA', 'MADRID', 'BUDAPEST', 'WIEN', 'LISBOA', 'PRAHA', 'BERLIN']


In [25]:
city_name_map = {
    "Athens": "ATHINA",
    "Milan": "MILANO",
    "Amsterdam": "AMSTERDAM",
    "Barcelona": "BARCELONA",
    "Madrid": "MADRID",
    "Budapest": "BUDAPEST",
    "Vienna": "WIEN",
    "Lisbon": "LISBOA",
    "Prague": "PRAHA",
    "Berlin": "BERLIN"
}

summer_lst_merge = summer_lst.copy()
summer_lst_merge["city"] = summer_lst_merge["city"].map(city_name_map)

lst_priority = summer_lst_merge.merge(
    final_df[
        ["city", "priority_score", "priority_rank", "uhi_intensity_c"]
    ],
    on="city",
    how="left"
)

print("Shape:", lst_priority.shape)

print("\nMissing after merge:")
print(
    lst_priority[
        ["priority_score", "priority_rank", "uhi_intensity_c"]
    ].isna().sum()
)

lst_priority.head(10)

Shape: (40, 7)

Missing after merge:
priority_score     0
priority_rank      0
uhi_intensity_c    0
dtype: int64


,city,year,summer_mean_lst_c,avg_pixel_count,priority_score,priority_rank,uhi_intensity_c
0,AMSTERDAM,2021,26.624178,3305.000000,0.588412,3,1.44617
1,AMSTERDAM,2022,29.644736,3354.000000,0.588412,3,1.44617
2,AMSTERDAM,2023,27.445384,3342.333333,0.588412,3,1.44617
3,AMSTERDAM,2024,27.227859,3346.666667,0.588412,3,1.44617
4,ATHINA,2021,40.030468,1926.333333,0.699785,1,2.10118
5,ATHINA,2022,38.977674,1919.000000,0.699785,1,2.10118
6,ATHINA,2023,40.663328,1920.666667,0.699785,1,2.10118
7,ATHINA,2024,42.215510,1928.000000,0.699785,1,2.10118
8,BARCELONA,2021,35.913541,2800.000000,0.509404,4,1.37022
9,BARCELONA,2022,38.225863,2806.333333,0.509404,4,1.37022


## LST, UHI and Cooling Priority Comparison

To evaluate whether the existing prioritisation is supported by the new heat data, each city's recent summer LST is compared with its UHI intensity and Priority Score.

The four annual summer LST values are first summarized into a city-level mean. This allows the temporal LST dataset to be compared consistently with the city-level baseline indicators.

The comparison is descriptive and is used to identify whether different measures of heat and vulnerability show similar or contrasting city patterns.

In [26]:
lst_city_summary = (
    lst_priority
    .groupby("city", as_index=False)
    .agg(
        mean_lst_2021_2024_c=("summer_mean_lst_c", "mean"),
        priority_score=("priority_score", "first"),
        priority_rank=("priority_rank", "first"),
        uhi_intensity_c=("uhi_intensity_c", "first")
    )
)

lst_city_summary = lst_city_summary.sort_values(
    "priority_rank"
)

lst_city_summary

,city,mean_lst_2021_2024_c,priority_score,priority_rank,uhi_intensity_c
1,ATHINA,40.471745,0.699785,1,2.10118
7,MILANO,35.188971,0.689442,2,2.00832
0,AMSTERDAM,27.735539,0.588412,3,1.44617
2,BARCELONA,36.744699,0.509404,4,1.37022
6,MADRID,43.025298,0.456055,5,1.72486
4,BUDAPEST,34.701179,0.389235,6,1.99352
9,WIEN,32.167619,0.384382,7,1.65995
5,LISBOA,37.886156,0.361484,8,1.11103
8,PRAHA,29.681964,0.331381,9,1.70673
3,BERLIN,29.720751,0.233595,10,1.73047


## Relationship Between LST, UHI and Priority Score

Pearson correlations are calculated to explore how the new LST measure relates to UHI intensity and the existing Priority Score.

This helps determine whether cities with higher average summer surface temperatures also tend to have stronger urban heat island intensity or higher overall cooling priority.

Because the analysis includes only 10 cities, the correlations are interpreted as descriptive associations rather than causal relationships.

In [27]:
lst_uhi_corr = lst_city_summary[
    "mean_lst_2021_2024_c"
].corr(
    lst_city_summary["uhi_intensity_c"]
)

lst_priority_corr = lst_city_summary[
    "mean_lst_2021_2024_c"
].corr(
    lst_city_summary["priority_score"]
)

print(f"LST vs UHI correlation: {lst_uhi_corr:.3f}")
print(f"LST vs Priority Score correlation: {lst_priority_corr:.3f}")

LST vs UHI correlation: 0.131
LST vs Priority Score correlation: 0.334


## Summer Heat Persistence and Variability

Beyond comparing individual years, the four summer observations are used to examine how persistent and variable surface heat is across the cities.

For each city, the analysis calculates the four-year mean summer LST, the minimum and maximum summer LST, and the range between them.

This helps distinguish cities that remain consistently hot from cities that experience stronger year-to-year fluctuations.

In [28]:
lst_patterns = (
    summer_lst
    .groupby("city", as_index=False)
    .agg(
        mean_lst_c=("summer_mean_lst_c", "mean"),
        min_lst_c=("summer_mean_lst_c", "min"),
        max_lst_c=("summer_mean_lst_c", "max"),
        lst_range_c=("summer_mean_lst_c", lambda x: x.max() - x.min())
    )
)

lst_patterns = lst_patterns.sort_values(
    "mean_lst_c",
    ascending=False
)

lst_patterns

,city,mean_lst_c,min_lst_c,max_lst_c,lst_range_c
6,Madrid,43.025298,41.543050,44.091009,2.547958
1,Athens,40.471745,38.977674,42.215510,3.237836
5,Lisbon,37.886156,36.671122,38.917975,2.246853
2,Barcelona,36.744699,35.530212,38.225863,2.695651
7,Milan,35.188971,34.144988,37.885319,3.740331
4,Budapest,34.701179,33.045675,37.580811,4.535135
9,Vienna,32.167619,31.192992,32.858674,1.665682
3,Berlin,29.720751,28.987234,30.920294,1.933060
8,Prague,29.681964,28.538186,30.843977,2.305791
0,Amsterdam,27.735539,26.624178,29.644736,3.020558


## Recent Urban Pressure and Surface Heat

Impervious-surface change from 2021 to 2024 is compared with recent summer LST patterns to add an urban-development perspective to the heat analysis.

The impervious increase indicator represents the percentage of valid pixels classified as increasing in imperviousness during 2021–2024.

This analysis explores whether cities experiencing higher or increasing surface heat are also showing recent increases in impervious surfaces. The relationship is interpreted descriptively and does not imply causation.

In [29]:
urban_pressure = lst_patterns.copy()

urban_pressure["city"] = urban_pressure["city"].map(city_name_map)

urban_pressure = urban_pressure.merge(
    final_df[
        ["city", "priority_score", "priority_rank", "impervious_increase_pct"]
    ],
    on="city",
    how="left"
)

urban_pressure = urban_pressure.merge(
    lst_change[["city", "change_2021_2024_c"]]
    .assign(city=lambda x: x["city"].map(city_name_map)),
    on="city",
    how="left"
)

urban_pressure = urban_pressure.sort_values(
    "impervious_increase_pct",
    ascending=False
)

urban_pressure

,city,mean_lst_c,min_lst_c,max_lst_c,lst_range_c,priority_score,priority_rank,impervious_increase_pct,change_2021_2024_c
4,MILANO,35.188971,34.144988,37.885319,3.740331,0.689442,2,0.43,0.335635
2,LISBOA,37.886156,36.671122,38.917975,2.246853,0.361484,8,0.35,0.498168
9,AMSTERDAM,27.735539,26.624178,29.644736,3.020558,0.588412,3,0.32,0.603682
0,MADRID,43.025298,41.543050,44.091009,2.547958,0.456055,5,0.29,1.183917
1,ATHINA,40.471745,38.977674,42.215510,3.237836,0.699785,1,0.28,2.185042
8,PRAHA,29.681964,28.538186,30.843977,2.305791,0.331381,9,0.26,2.258315
6,WIEN,32.167619,31.192992,32.858674,1.665682,0.384382,7,0.21,1.665682
3,BARCELONA,36.744699,35.530212,38.225863,2.695651,0.509404,4,0.20,-0.383329
5,BUDAPEST,34.701179,33.045675,37.580811,4.535135,0.389235,6,0.19,1.646503
7,BERLIN,29.720751,28.987234,30.920294,1.933060,0.233595,10,0.19,0.141584


## Relationship Between Recent Urban Pressure and LST Change

To examine whether recent increases in impervious surfaces are associated with changes in summer surface temperature, the 2021–2024 impervious increase indicator is compared with the change in summer mean LST over the same period.

Because the analysis includes only 10 cities and covers a short four-year period, the result is interpreted as a descriptive association rather than evidence of causation.

In [30]:
impervious_lst_corr = urban_pressure[
    "impervious_increase_pct"
].corr(
    urban_pressure["change_2021_2024_c"]
)

print(
    f"Impervious Increase vs LST Change correlation: "
    f"{impervious_lst_corr:.3f}"
)

Impervious Increase vs LST Change correlation: -0.139


## Priority Score Sensitivity Analysis

The baseline Priority Score gives equal weight to all six indicators. To test whether the city ranking depends strongly on this assumption, alternative weighting scenarios are evaluated.

Three alternative scenarios are compared with the equal-weight baseline:

- **Heat-focused:** greater emphasis on UHI intensity.
- **Exposure-focused:** greater emphasis on population density.
- **Cooling-focused:** greater emphasis on green-space and street-tree deficits.

The purpose is not to identify one “correct” weighting scheme, but to evaluate how stable the city priorities remain under different planning perspectives.

In [31]:
score_cols = {
    "building": "building_density_per_km2_score",
    "height": "average_building_height_m_score",
    "trees": "street_tree_pct_score",
    "green": "green_space_pct_score",
    "uhi": "uhi_intensity_c_score",
    "population": "population_density_per_km2_score"
}

# Alternative weighting scenarios
weights = {
    "equal": {
        "building": 1/6, "height": 1/6, "trees": 1/6,
        "green": 1/6, "uhi": 1/6, "population": 1/6
    },
    "heat_focused": {
        "building": 0.10, "height": 0.10, "trees": 0.10,
        "green": 0.10, "uhi": 0.40, "population": 0.20
    },
    "exposure_focused": {
        "building": 0.10, "height": 0.10, "trees": 0.10,
        "green": 0.10, "uhi": 0.20, "population": 0.40
    },
    "cooling_focused": {
        "building": 0.10, "height": 0.10, "trees": 0.25,
        "green": 0.25, "uhi": 0.15, "population": 0.15
    }
}

sensitivity = final_df.copy()

for scenario, w in weights.items():
    sensitivity[f"{scenario}_score"] = sum(
        sensitivity[score_cols[k]] * weight
        for k, weight in w.items()
    )

sensitivity[
    ["city", "priority_score", "equal_score",
     "heat_focused_score", "exposure_focused_score",
     "cooling_focused_score"]
].round(3)

,city,priority_score,equal_score,heat_focused_score,exposure_focused_score,cooling_focused_score
0,ATHINA,0.700,0.700,0.816,0.807,0.619
1,MILANO,0.689,0.689,0.765,0.742,0.681
2,AMSTERDAM,0.588,0.588,0.489,0.490,0.598
3,BARCELONA,0.509,0.509,0.484,0.632,0.445
4,MADRID,0.456,0.456,0.495,0.442,0.502
5,BUDAPEST,0.389,0.389,0.511,0.352,0.432
6,WIEN,0.384,0.384,0.397,0.286,0.498
7,LISBOA,0.361,0.361,0.240,0.285,0.422
8,PRAHA,0.331,0.331,0.384,0.273,0.397
9,BERLIN,0.234,0.234,0.342,0.245,0.209


In [32]:
scenarios = [
    "equal",
    "heat_focused",
    "exposure_focused",
    "cooling_focused"
]

for scenario in scenarios:
    sensitivity[f"{scenario}_rank"] = (
        sensitivity[f"{scenario}_score"]
        .rank(method="min", ascending=False)
        .astype(int)
    )

rank_comparison = sensitivity[
    [
        "city",
        "equal_rank",
        "heat_focused_rank",
        "exposure_focused_rank",
        "cooling_focused_rank"
    ]
].sort_values("equal_rank")

rank_comparison

,city,equal_rank,heat_focused_rank,exposure_focused_rank,cooling_focused_rank
0,ATHINA,1,1,1,2
1,MILANO,2,2,2,1
2,AMSTERDAM,3,5,4,3
3,BARCELONA,4,6,3,6
4,MADRID,5,4,5,4
5,BUDAPEST,6,3,6,7
6,WIEN,7,7,7,5
7,LISBOA,8,10,8,8
8,PRAHA,9,8,9,9
9,BERLIN,10,9,10,10


## Priority Rank Stability

Rank stability measures how much each city's position changes across the alternative weighting scenarios.

The minimum and maximum rank are compared to calculate the total rank range. A smaller range indicates a more stable priority classification, while a larger range indicates greater sensitivity to the chosen planning perspective.

In [33]:
rank_cols = [
    "equal_rank",
    "heat_focused_rank",
    "exposure_focused_rank",
    "cooling_focused_rank"
]

rank_stability = rank_comparison.copy()

rank_stability["best_rank"] = rank_stability[rank_cols].min(axis=1)
rank_stability["worst_rank"] = rank_stability[rank_cols].max(axis=1)

rank_stability["rank_range"] = (
    rank_stability["worst_rank"] -
    rank_stability["best_rank"]
)

rank_stability = rank_stability.sort_values(
    ["rank_range", "equal_rank"]
)

rank_stability

,city,equal_rank,heat_focused_rank,exposure_focused_rank,cooling_focused_rank,best_rank,worst_rank,rank_range
0,ATHINA,1,1,1,2,1,2,1
1,MILANO,2,2,2,1,1,2,1
4,MADRID,5,4,5,4,4,5,1
8,PRAHA,9,8,9,9,8,9,1
9,BERLIN,10,9,10,10,9,10,1
2,AMSTERDAM,3,5,4,3,3,5,2
6,WIEN,7,7,7,5,5,7,2
7,LISBOA,8,10,8,8,8,10,2
3,BARCELONA,4,6,3,6,3,6,3
5,BUDAPEST,6,3,6,7,3,7,4


## Exporting Analytical Results

The final outputs from the temporal LST analysis and Priority Score sensitivity analysis are exported as separate analytical datasets for later use in SQL and visualisation.

In [34]:
output_dir = Path("../Data_Processed")
output_dir.mkdir(parents=True, exist_ok=True)

summer_lst.to_csv(
    output_dir / "LST_Summer_2021_2024.csv",
    index=False
)

rank_stability.to_csv(
    output_dir / "Priority_Sensitivity_Analysis.csv",
    index=False
)

print("Files saved successfully.")

Files saved successfully.


## Intervention Gap Analysis

To strengthen the link between the vulnerability assessment and recommended cooling actions, intervention gaps are calculated for green-space and street-tree coverage.

The median coverage across the 10 study cities is used as a transparent internal benchmark. For cities below the benchmark, the analysis measures how much additional green-space or street-tree coverage would be required to reach the study median.

These gaps are used as decision-support indicators rather than fixed policy targets.

In [35]:
green_benchmark = final_df["green_space_pct"].median()
tree_benchmark = final_df["street_tree_pct"].median()

intervention_gap = final_df[
    [
        "city",
        "priority_rank",
        "green_space_pct",
        "street_tree_pct",
        "recommended_intervention"
    ]
].copy()

intervention_gap["green_space_gap_pct"] = (
    green_benchmark - intervention_gap["green_space_pct"]
).clip(lower=0)

intervention_gap["street_tree_gap_pct"] = (
    tree_benchmark - intervention_gap["street_tree_pct"]
).clip(lower=0)

print(f"Green-space benchmark: {green_benchmark:.2f}%")
print(f"Street-tree benchmark: {tree_benchmark:.2f}%")

intervention_gap.sort_values("priority_rank")

Green-space benchmark: 32.44%
Street-tree benchmark: 3.61%


,city,priority_rank,green_space_pct,street_tree_pct,recommended_intervention,green_space_gap_pct,street_tree_gap_pct
0,ATHINA,1,47.222665,3.254409,Expand targeted cooling and shade,0.000000,0.358227
1,MILANO,2,9.927272,4.814992,Expand green infrastructure,22.515454,0.000000
2,AMSTERDAM,3,8.189335,4.219192,Expand green infrastructure,24.253391,0.000000
3,BARCELONA,4,48.296473,3.817476,Target cooling in high-density areas,0.000000,0.000000
4,MADRID,5,37.972160,2.091410,Expand street-tree coverage,0.000000,1.521226
5,BUDAPEST,6,24.274084,4.261056,Strengthen urban cooling and tree canopy,8.168642,0.000000
6,WIEN,7,23.588846,1.929496,Expand green infrastructure and street-tree co...,8.853880,1.683141
7,LISBOA,8,36.514981,1.846103,Expand street-tree coverage,0.000000,1.766533
8,PRAHA,9,28.370472,3.407796,Strengthen green infrastructure and shade,4.072255,0.204840
9,BERLIN,10,40.156175,5.840580,Target cooling in dense built-up areas,0.000000,0.000000


In [36]:
intervention_gap.to_csv(
    output_dir / "Intervention_Gap_Analysis.csv",
    index=False
)

print("Intervention Gap saved successfully.")

Intervention Gap saved successfully.


## Final Recommendation Review

The existing city recommendations are reviewed against the Priority Score drivers, recent summer LST patterns, and the quantified intervention gaps.

Recommendations are only revised where the combined analytical evidence indicates that a different intervention would better address the city's identified vulnerability profile.

In [38]:
print(lst_patterns.columns.tolist())

['city', 'mean_lst_c', 'min_lst_c', 'max_lst_c', 'lst_range_c']


In [39]:
recommendation_review = (
    final_df[
        [
            "city",
            "priority_rank",
            "primary_driver",
            "secondary_driver",
            "recommended_intervention"
        ]
    ]
    .merge(
        lst_patterns[["city", "mean_lst_c", "lst_range_c"]],
        on="city",
        how="left"
    )
    .merge(
        intervention_gap[
            ["city", "green_space_gap_pct", "street_tree_gap_pct"]
        ],
        on="city",
        how="left"
    )
    .sort_values("priority_rank")
)

recommendation_review

,city,priority_rank,primary_driver,secondary_driver,recommended_intervention,mean_lst_c,lst_range_c,green_space_gap_pct,street_tree_gap_pct
0,ATHINA,1,UHI Intensity,Population Density,Expand targeted cooling and shade,NaN,NaN,0.000000,0.358227
1,MILANO,2,Green-Space Deficit,UHI Intensity,Expand green infrastructure,NaN,NaN,22.515454,0.000000
2,AMSTERDAM,3,Building Density,Green-Space Deficit,Expand green infrastructure,NaN,NaN,24.253391,0.000000
3,BARCELONA,4,Building Height,Population Density,Target cooling in high-density areas,NaN,NaN,0.000000,0.000000
4,MADRID,5,Street-Tree Deficit,UHI Intensity,Expand street-tree coverage,NaN,NaN,0.000000,1.521226
5,BUDAPEST,6,UHI Intensity,Green-Space Deficit,Strengthen urban cooling and tree canopy,NaN,NaN,8.168642,0.000000
6,WIEN,7,Street-Tree Deficit,Green-Space Deficit,Expand green infrastructure and street-tree co...,NaN,NaN,8.853880,1.683141
7,LISBOA,8,Street-Tree Deficit,Building Height,Expand street-tree coverage,NaN,NaN,0.000000,1.766533
8,PRAHA,9,Street-Tree Deficit,UHI Intensity,Strengthen green infrastructure and shade,NaN,NaN,4.072255,0.204840
9,BERLIN,10,UHI Intensity,Building Density,Target cooling in dense built-up areas,NaN,NaN,0.000000,0.000000


In [40]:
lst_patterns_fixed = lst_patterns.copy()

lst_patterns_fixed["city"] = lst_patterns_fixed["city"].map(city_name_map)

recommendation_review = (
    final_df[
        [
            "city",
            "priority_rank",
            "primary_driver",
            "secondary_driver",
            "recommended_intervention"
        ]
    ]
    .merge(
        lst_patterns_fixed[["city", "mean_lst_c", "lst_range_c"]],
        on="city",
        how="left"
    )
    .merge(
        intervention_gap[
            ["city", "green_space_gap_pct", "street_tree_gap_pct"]
        ],
        on="city",
        how="left"
    )
    .sort_values("priority_rank")
)

recommendation_review

,city,priority_rank,primary_driver,secondary_driver,recommended_intervention,mean_lst_c,lst_range_c,green_space_gap_pct,street_tree_gap_pct
0,ATHINA,1,UHI Intensity,Population Density,Expand targeted cooling and shade,40.471745,3.237836,0.000000,0.358227
1,MILANO,2,Green-Space Deficit,UHI Intensity,Expand green infrastructure,35.188971,3.740331,22.515454,0.000000
2,AMSTERDAM,3,Building Density,Green-Space Deficit,Expand green infrastructure,27.735539,3.020558,24.253391,0.000000
3,BARCELONA,4,Building Height,Population Density,Target cooling in high-density areas,36.744699,2.695651,0.000000,0.000000
4,MADRID,5,Street-Tree Deficit,UHI Intensity,Expand street-tree coverage,43.025298,2.547958,0.000000,1.521226
5,BUDAPEST,6,UHI Intensity,Green-Space Deficit,Strengthen urban cooling and tree canopy,34.701179,4.535135,8.168642,0.000000
6,WIEN,7,Street-Tree Deficit,Green-Space Deficit,Expand green infrastructure and street-tree co...,32.167619,1.665682,8.853880,1.683141
7,LISBOA,8,Street-Tree Deficit,Building Height,Expand street-tree coverage,37.886156,2.246853,0.000000,1.766533
8,PRAHA,9,Street-Tree Deficit,UHI Intensity,Strengthen green infrastructure and shade,29.681964,2.305791,4.072255,0.204840
9,BERLIN,10,UHI Intensity,Building Density,Target cooling in dense built-up areas,29.720751,1.933060,0.000000,0.000000


### Key Findings

- Summer surface temperatures varied considerably across the 10 cities between 2021 and 2024.
- Madrid recorded the highest four-year mean summer LST (43.03°C), followed by Athens (40.47°C).
- Nine of the ten cities had a higher summer mean LST in 2024 than in 2021; Barcelona was the only exception.
- Budapest showed the highest year-to-year variability, with an LST range of 4.54°C.
- The four-year period is interpreted as a recent temporal comparison rather than a long-term climate trend.

### Key Findings

- LST and UHI show only a very weak positive correlation (r = 0.131).
- LST and Priority Score show a weak-to-moderate positive correlation (r = 0.334).
- High surface temperature alone does not determine a city's cooling priority.
- Madrid has the highest four-year mean LST but ranks fifth in overall cooling priority.
- The results support using a multidimensional framework that combines heat hazard, exposure, urban form and cooling capacity.

### Key Findings

- Recent impervious-surface increase shows almost no relationship with the 2021–2024 change in summer LST (r = -0.139).
- Milan recorded the highest share of pixels with increased imperviousness (0.43%), followed by Lisbon (0.35%) and Amsterdam (0.32%).
- Cities with larger recent impervious increases did not consistently show larger LST increases.
- Recent impervious change therefore provides additional urban-development context but does not explain short-term LST differences on its own.

### Key Findings

- Athens and Milan remain the two highest-priority cities across all weighting scenarios, indicating robust high-priority classifications.
- Berlin remains consistently among the lowest-priority cities.
- Madrid also shows a stable ranking, remaining between fourth and fifth place.
- Budapest is the most sensitive city, ranging from third to seventh depending on the weighting scenario.
- Barcelona also shows moderate sensitivity, ranging from third to sixth.
- Overall, the highest priorities remain stable, while some mid-ranked cities are more dependent on the planning perspective used.

### Key Findings

- Milan and Amsterdam show the largest green-space gaps, supporting green-infrastructure interventions.
- Madrid and Lisbon show substantial street-tree gaps, supporting targeted expansion of street-tree coverage.
- Vienna has deficits in both green-space and street-tree coverage, supporting a combined intervention.
- Barcelona and Berlin are already above the study median for both indicators, suggesting that their cooling priorities are driven by other factors.
- Budapest shows a green-space gap but no street-tree gap, indicating that its recommendation should focus more on green infrastructure and targeted cooling.

# Final Analytical Summary

The extended analysis strengthens the urban cooling prioritisation framework by adding temporal heat patterns, recent urban-development context, model robustness testing, and quantified intervention gaps.

### Main Findings

- **Athens and Milan are robust high-priority cities**, remaining first or second across all weighting scenarios.
- **Surface heat alone does not determine cooling priority.** Madrid records the highest four-year mean summer LST, but ranks fifth in the multidimensional Priority Score.
- **Recent summer heat patterns differ substantially across cities.** Nine of the ten cities recorded higher summer mean LST in 2024 than in 2021, while Budapest showed the greatest year-to-year variability.
- **Recent impervious-surface increase alone does not explain LST change** across the 10-city sample.
- **Some mid-ranked cities are sensitive to planning priorities.** Budapest shows the largest ranking variation across alternative weighting scenarios.
- **Intervention needs differ by city.** Green-space gaps are particularly large in Amsterdam and Milan, while street-tree gaps are more important in cities such as Madrid and Lisbon.
- The results support a **city-specific, multidimensional approach** to urban cooling rather than prioritising cities based on a single heat indicator.